From e1777741dd3e3787a805461179edeb9c51dcca26 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Mon, 11 Jan 2021 12:05:31 +0200
Subject: [PATCH] [SW-0] bitops: fix the __ffs32() version

The commit
b101837 ("[SW-0] bitops: introduce dedicated header")
has chosen the wrong __ffs32() version which should be used differently
than our regular code expects.
Fix this by introducing the correct (expected) version.

Fixes: b101837b27cfdc95cad4fb65e4a63388189d7f68

Change-Id: If88e16a3fbbe23ce7971ec0900ba059f8a5d7c05
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/include/bitops.h b/include/bitops.h
index ba2d559..5737a58 100644
--- a/include/bitops.h
+++ b/include/bitops.h
@@ -15,40 +15,35 @@
 
 #ifndef __ASSEMBLY__
 
-/*
- * __ffs32: find first bit set. This is defined the same way as
- * the libc and compiler builtin ffs routines, therefore
- * differs in spirit from the above ffz (man ffs).
+/**
+ * __ffs - find first bit in word.
+ * @word: The word to search
+ *
+ * Undefined if no bit exists, so code should check against 0 first.
  */
-
-static inline int __ffs32(unsigned int x)
+static inline int __ffs32(unsigned int word)
 {
-	int r = 1;
+	int num = 0;
 
-	if (!x)
-		return 0;
-
-	if (!(x & 0xffff)) {
-		x >>= 16;
-		r += 16;
+	if ((word & 0xffff) == 0) {
+		num += 16;
+		word >>= 16;
 	}
-	if (!(x & 0xff)) {
-		x >>= 8;
-		r += 8;
+	if ((word & 0xff) == 0) {
+		num += 8;
+		word >>= 8;
 	}
-	if (!(x & 0xf)) {
-		x >>= 4;
-		r += 4;
+	if ((word & 0xf) == 0) {
+		num += 4;
+		word >>= 4;
 	}
-	if (!(x & 3)) {
-		x >>= 2;
-		r += 2;
+	if ((word & 0x3) == 0) {
+		num += 2;
+		word >>= 2;
 	}
-	if (!(x & 1)) {
-		x >>= 1;
-		r += 1;
-	}
-	return r;
+	if ((word & 0x1) == 0)
+		num += 1;
+	return num;
 }
 
 #endif /* __ASSEMBLY__ */
